symengine/tests/eval/test_evalf.cpp

```diff
@@ -0,0 +1,170 @@
+
+using SymEngine::Basic;
+using SymEngine::constant;
+using SymEngine::Complex;
+using SymEngine::RealDouble;
+using SymEngine::ComplexDouble;
+using SymEngine::complex_double;
+using SymEngine::real_double;
+#ifdef HAVE_SYMENGINE_MPFR
+using SymEngine::real_mpfr;
+using SymEngine::RealMPFR;
+#endif // HAVE_SYMENGINE_MPFR
+#ifdef HAVE_SYMENGINE_MPC
+using SymEngine::complex_mpc;
+using SymEngine::ComplexMPC;
+#endif // HAVE_SYMENGINE_MPC
+using SymEngine::Rational;
+using SymEngine::symbol;
+using SymEngine::integer;
+using SymEngine::integer_class;
+using SymEngine::one;
+using SymEngine::zero;
+using SymEngine::pow;
+using SymEngine::RCP;
+using SymEngine::print_stack_on_segfault;
+using SymEngine::levi_civita;
+using SymEngine::zeta;
+using SymEngine::sin;
+using SymEngine::cos;
+using SymEngine::tan;
+using SymEngine::sec;
+using SymEngine::csc;
+using SymEngine::cot;
+using SymEngine::asin;
+using SymEngine::acos;
+using SymEngine::atan;
+using SymEngine::asec;
+using SymEngine::acsc;
+using SymEngine::acot;
+using SymEngine::sinh;
+using SymEngine::cosh;
+using SymEngine::tanh;
+using SymEngine::coth;
+using SymEngine::asinh;
+using SymEngine::acosh;
+using SymEngine::atanh;
+using SymEngine::acoth;
+using SymEngine::erf;
+using SymEngine::log;
+using SymEngine::pi;
+using SymEngine::I;
+using SymEngine::E;
+using SymEngine::EulerGamma;
+using SymEngine::loggamma;
+using SymEngine::vec_basic;
+using SymEngine::rational_class;
+using SymEngine::max;
+using SymEngine::min;
+using SymEngine::min;
+using SymEngine::rcp_static_cast;
+
+TEST_CASE("evalf: real_double", "[evalf]")
+{
+    RCP<const Basic> r1, r2;
+    r1 = sin(integer(2));
+    r2 = evalf(*r1, 53, true);
+    // SYMENGINE_REAL_DOUBLE = 6
+    REQUIRE(r2->get_type_code() == 6);
+    double d1 = 0.909297;
+    double d2 = (rcp_static_cast<const RealDouble>(r2))->as_double();
+    d1 = fabs(d1 - d2);
+    d2 = 0.000001;
+    REQUIRE(d1 < d2);
+}
+
+#ifdef HAVE_SYMENGINE_MPFR
+TEST_CASE("evalf: real_mpfr", "[evalf]")
+{
+    RCP<const Basic> r1, r2;
+    r1 = mul(pi, integer(integer_class("1963319607")));
+    r2 = integer(integer_class("6167950454"));
+    r1 = sub(r1, r2);
+
+    r2 = evalf(*r1, 100, true);
+    // SYMENGINE_REAL_MPFR = 4
+    REQUIRE(r2->get_type_code() == 4);
+    REQUIRE(!(rcp_static_cast<const RealMPFR>(r2))->is_zero());
+
+    r2 = evalf(*r1, 60, true);
+    // SYMENGINE_REAL_MPFR = 4
+    REQUIRE(r2->get_type_code() == 4);
+    REQUIRE((rcp_static_cast<const RealMPFR>(r2))->is_zero());
+}
+#endif // HAVE_SYMENGINE_MPFR
+
+TEST_CASE("evalf: complex_double", "[evalf]")
+{
+    RCP<const Basic> r1, r2;
+    r1 = add(sin(integer(4)), mul(sin(integer(3)), I));
+    // r1 = sin(4) + sin(3)i
+    r2 = add(sin(integer(2)), mul(sin(integer(7)), I));
+    // r2 = sin(2) + sin(7)i
+
+    r1 = mul(r1, r2);
+    // r1 = (sin(4) + sin(3)i) * (sin(2) + sin(7)i)
+
+    r2 = evalf(*r1, 53, false);
+    // SYMENGINE_COMPLEX_DOUBLE = 3
+    REQUIRE(r2->get_type_code() == 3);
```

Comment:
Try to use the actual enums(I am sure there should be a way) everywhere you have written like this, because if any new types are added in between these tests will fail.